testsuite/src/main/java/io/netty/testsuite/transport/socket/AbstractSocketShutdownOutputByPeerTest.java

```diff
@@ -0,0 +1,162 @@
+import io.netty.channel.ChannelOption;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.channel.socket.ChannelInputShutdownEvent;
+import io.netty.channel.socket.DuplexChannel;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.*;
+
+public abstract class AbstractSocketShutdownOutputByPeerTest<Socket> extends AbstractServerSocketTest {
+
+    @Test(timeout = 30000)
+    public void testShutdownOutput() throws Throwable {
+        run();
+    }
+
+    public void testShutdownOutput(ServerBootstrap sb) throws Throwable {
+        TestHandler h = new TestHandler();
+        Socket s = newSocket();
+        Channel sc = null;
+        try {
+            sc = sb.childHandler(h).childOption(ChannelOption.ALLOW_HALF_CLOSURE, true).bind().sync().channel();
+
+            connect(s, sc.localAddress());
+            write(s, 1);
+
+            assertEquals(1, (int) h.queue.take());
+
+            assertTrue(h.ch.isOpen());
+            assertTrue(h.ch.isActive());
+            assertFalse(h.ch.isInputShutdown());
+            assertFalse(h.ch.isOutputShutdown());
+
+            shutdownOutput(s);
+
+            h.halfClosure.await();
+
+            assertTrue(h.ch.isOpen());
+            assertTrue(h.ch.isActive());
+            assertTrue(h.ch.isInputShutdown());
+            assertFalse(h.ch.isOutputShutdown());
+            assertEquals(1, h.closure.getCount());
+            Thread.sleep(100);
+            assertEquals(1, h.halfClosureCount.intValue());
+        } finally {
+            if (sc != null) {
+                sc.close();
+            }
+            close(s);
+        }
+    }
+
+    @Test(timeout = 30000)
+    public void testShutdownOutputWithoutOption() throws Throwable {
+        run();
+    }
+
+    public void testShutdownOutputWithoutOption(ServerBootstrap sb) throws Throwable {
+        TestHandler h = new TestHandler();
+        Socket s = newSocket();
+        Channel sc = null;
+        try {
+            sc = sb.childHandler(h).bind().sync().channel();
+
+            connect(s, sc.localAddress());
+            write(s, 1);
+
+            assertEquals(1, (int) h.queue.take());
+
+            assertTrue(h.ch.isOpen());
+            assertTrue(h.ch.isActive());
+            assertFalse(h.ch.isInputShutdown());
+            assertFalse(h.ch.isOutputShutdown());
+
+            shutdownOutput(s);
+
+            h.closure.await();
+
+            assertFalse(h.ch.isOpen());
+            assertFalse(h.ch.isActive());
+            assertTrue(h.ch.isInputShutdown());
+            assertTrue(h.ch.isOutputShutdown());
+
+            assertEquals(1, h.halfClosure.getCount());
+            Thread.sleep(100);
+            assertEquals(0, h.halfClosureCount.intValue());
```

Comment:
@zhongl I wonder if it may be better to do something like this to reduce the risk of false-positives:

```
while (h.halfClosure.getCount() != 1 && h.halfClosureCount.intValue() != 0) {
    Thread.sleep(100);
}
```

Just ensure you use a timeout in the `@Test` declaration so we not end up in an endless loop if something went wrong

Reply:
ok, how about using this while pattern to improve another thread sleeping at line 70?

Reply:
sounds good

Reply:
done!